### ISSUE TITLE
Coding standard should forbid the use of "foo %s" % notATuple

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3045]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)| @dreid reported|
|-|-|
|Trac ID|trac#3045|
|Type|enhancement|
|Created|2008-02-20 21:43:56Z|
|Branch|https://github.com/twisted/twisted/tree/string-format-standard-3045|

Usage of this has been discourage for some time however it should probably be listed in the coding standard.

As exarkun points out it's also OK for the right side of the % operator to be a dict.

<details><summary>Searchable metadata</summary>

```
trac-id__3045 3045
type__enhancement enhancement
reporter__dreid dreid
priority__low low
milestone__ 
branch__branches_string_format_standard_3045 branches/string-format-standard-3045
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1203543836000000 1203543836000000
changetime__1220914460000000 1220914460000000
version__None None
owner__ 
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_2"></a>|@thijstriemstra commented|
|-|-|

(In [24681]) Branching to 'string-format-standard-3045'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Added a 'string formatting section' to the coding standard in r24682. Putting it up for review.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

From today's IRC chat:

```
radix: thijs: your changes about string formatting in the coding standard don't describe our standard, they describe python's semantics
[15:10] radix: aiui, our standard is that you should *always* use a tuple if you're using non-key-based replacement tokens
[15:10] thijs: radix: true, intended as a start, so you can review it and give me the actual coding standard
[15:10] thijs:
[15:10] radix: ah ok 
[15:11] radix: thijs: ok, so even if you have a single %s, you should use a single-item tuple
[15:12] thijs: ok
[15:12] radix: this is to avoid unexpected behavior when you think you're passing in a single value but the value is unexpectedly a tuple
[15:12] radix: e.g.
[15:12] radix: def foo(x): return "Hi %s\n" % x
[15:12] radix: that's bad
[15:13] radix: you can pass in foo("foo") or foo(3) fine, but if you pass in foo((1,2)), it blows up
[15:13] thijs: radix: thx, will make the changes
[15:13] radix: if you make it return "Hi %s\n" % (x,), it always works
```

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Added radix's notes in r24699, putting it up for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

There are no more acceptance tests.  Can you file a ticket for deleting the part of the coding standard which discusses them?  Since this will be deleted, there's probably no reason to change _nonportable_ to _non-portable_.

_whitespace_ is an acceptable spelling, the hyphenated _white-space_ is much less common.  _whitespaces_ (plural) is weird though, so if you want to drop that _s_, that's fine.

The 2nd line of the _String Formatting Operations_ paragraph is long, can you reflow that paragraph?

Can you replace _blows up_ with _raises a TypeError_?

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra set status to `assigned`|
|-|-|

Replying to [exarkun](#note_7):
> There are no more acceptance tests.  Can you file a ticket for deleting the part of the coding standard which discusses them?  Since this will be deleted, there's probably no reason to change _nonportable_ to _non-portable_.

Opened [#3422](https://github.com/twisted/twisted/issues/3422)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Replying to [exarkun](#note_7):
> Since this will be deleted, there's probably no reason to change _nonportable_ to _non-portable_.

Reverted this change.

> _whitespace_ is an acceptable spelling, the hyphenated _white-space_ is much less common.  _whitespaces_ (plural) is weird though, so if you want to drop that _s_, that's fine.

Changed to 'whitespace'.

> The 2nd line of the _String Formatting Operations_ paragraph is long, can you reflow that paragraph?

I removed a redundant sentence.

> Can you replace _blows up_ with _raises a TypeError_?

fixed, see r24731 for changes.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [24751]) M-q

Reflow the paragraph so none of the lines are wider than 79 cols

refs [#3045](https://github.com/twisted/twisted/issues/3045)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [24752]) try to simplify language about non-mapping case

refs [#3045](https://github.com/twisted/twisted/issues/3045)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [24753]) Fix markup error

refs [#3045](https://github.com/twisted/twisted/issues/3045)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

Some edits, as visible above.  I think it looks good now.  If you agree, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_14"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [24754]) Merge string-format-standard-3045: Added a 'string formatting' section to the coding standard document.

Author: thijs, exarkun
Reviewer: radix
Fixes: [#3045](https://github.com/twisted/twisted/issues/3045)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
